lock-impl/src/main/java/com/palantir/lock/impl/HeldLocks.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright 2017 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock.impl;
+
+import javax.annotation.concurrent.Immutable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.palantir.lock.ExpiringToken;
+import com.palantir.lock.HeldLocksGrant;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockCollection;
+
+/**
+ * A set of locks held by the lock server, along with the canonical
+ * {@link HeldLocksToken} or {@link HeldLocksGrant} object for these locks.
```

Comment:
Should this be an immutables generated holder? Should define behavior of equals/hasCode in relation to token & collection